### PR TITLE
Corrected Boolean Expression StrPrinting except Not

### LINKED ISSUE #11435: str printing of logic expressions should use operators
Currently:

```
In [5]: print(And(Not(x), Or(y, z)))
And(Not(x), Or(y, z))
```

But it would be better for the str printer (and non-Unicode pretty printer) to use `~`, `&`, and `|`, like `~x & (y | z)`. 

This should be easy to fix, although you need to make sure the parenthesization is correct.

### PR DESCRIPTION
**Pull request by sagar-bharadwaj-ks**
Fixes #11435

### CODE REVIEW

File: sympy/printing/str.py [hunk 1]
```diff
@@ -74,12 +74,15 @@ def _print_BooleanTrue(self, expr):
     def _print_BooleanFalse(self, expr):
         return "False"
 
+	def _print_Not(self, expr):
+		return '~\(' + expr.args[0] + '\)'
```
Comment (Upabjojr) [inline sympy/printing/str.py]: maybe expressions should be parenthesized based on the node precedence.

Try using 
https://github.com/SagarB-97/sympy/blob/9f3f6b99c729d243060c4f5f0f49b9452bfbc972/sympy/printing/str.py#L27

File: sympy/printing/str.py [hunk 2]
```diff
@@ -74,12 +74,15 @@ def _print_BooleanTrue(self, expr):
     def _print_BooleanFalse(self, expr):
         return "False"
 
+	def _print_Not(self, expr):
+		return '~\(' + expr.args[0] + '\)' 
+
     def _print_And(self, expr):
-        return '%s(%s)' % (expr.func, ', '.join(sorted(self._print(a) for a in
+        return '(%s)' % (' & '.join(sorted(self._print(a) for a in
```
Comment (Upabjojr) [inline sympy/printing/str.py]: parenthesize needed.

File: sympy/printing/str.py [hunk 3]
```diff
@@ -74,12 +74,15 @@ def _print_BooleanTrue(self, expr):
...
             expr.args)))
 
     def _print_Or(self, expr):
-        return '%s(%s)' % (expr.func, ', '.join(sorted(self._print(a) for a in
+        return '(%s)' % (' | '.join(sorted(self._print(a) for a in
```
Comment (Upabjojr) [inline sympy/printing/str.py]: parenthesize needed.

File: sympy/printing/tests/test_str.py [hunk 4]
```diff
@@ -666,14 +666,14 @@ def test_settings():
...
+    assert str(where(X > 0)) == "Domain: (0 < x1 & x1 < oo)"
 
     D = Die('d1', 6)
-    assert str(where(D > 4)) == "Domain: Or(Eq(d1, 5), Eq(d1, 6))"
+    assert str(where(D > 4)) == "Domain: (Eq(d1, 5) | Eq(d1, 6))"
 
     A = Exponential('a', 1)
     B = Exponential('b', 1)
-    assert str(pspace(Tuple(A, B)).domain) == "Domain: And(0 <= a, 0 <= b, a < oo, b < oo)"
+    assert str(pspace(Tuple(A, B)).domain) == "Domain: (0 <= a & 0 <= b & a < oo & b < oo)"
```
Comment (Upabjojr) [inline sympy/printing/tests/test_str.py]: These are wrong, inequality operators have a lower precedence than `&`, they need parentheses:
`(0 <= a) & (0 <= b) & (a < oo) & (b < oo)`

File: sympy/printing/tests/test_str.py [hunk 5]
```diff
@@ -666,14 +666,14 @@ def test_settings():
 def test_RandomDomain():
     from sympy.stats import Normal, Die, Exponential, pspace, where
     X = Normal('x1', 0, 1)
-    assert str(where(X > 0)) == "Domain: And(0 < x1, x1 < oo)"
+    assert str(where(X > 0)) == "Domain: (0 < x1 & x1 < oo)"
```
Comment (Upabjojr) [inline sympy/printing/tests/test_str.py]: parentheses

File: sympy/printing/str.py (hunk 1 above)
Comment (SagarB-97) [inline sympy/printing/str.py]: @Upabjojr 
Thanks a Lot! I'll do that.
Can you tell me why Travis CI build is failing??
Also why is _print_Not() not getting called for str(Not(x))? (My guess is _print_Basic() is getting called)

**Comment by Upabjojr** [comment]
> Can you tell me why Travis CI build is failing??

* Do not use TABs (4 space)
* run `test("sympy/utilities")` to check code format.
* correct the parenthesis bug.
* check code printer tests.

File: sympy/printing/tests/test_str.py (hunk 4 above)
Comment (SagarB-97) [inline sympy/printing/tests/test_str.py]: @Upabjojr 
`&` is Bitwise AND in python and has  higher precedence than relational operators.
Shall I replace it with `and` which is python's Boolean AND and has lower precedence than relational operators?

I will use parenthesize() method anyway though!

Comment (Upabjojr) [inline sympy/printing/tests/test_str.py]: > & is Bitwise AND in python and has higher precedence than relational operators.

Yes, that's the problem. You need parentheses.

> Shall I replace it with and which is python's Boolean AND and has lower precedence than relational operators?

No, `and` is not supposed to return `And` objects, it should be evaluated and immediately return a result. `&` is supposed to return unevaluated `And` expressions in case the truthness cannot be immediately determined.

**Comment by Upabjojr** [comment]
> Also why is _print_Not() not getting called for str(Not(x))? (My guess is _print_Basic() is getting called)

Use a debugger to follow the calling tree, that's the way you get to know this kind of information. I personally use PyCharm.

File: sympy/printing/tests/test_str.py (hunk 4 above)
Comment (SagarB-97) [inline sympy/printing/tests/test_str.py]: Oh Okay! Thanks a lot. I get it now!!

**Comment by SagarB-97** [comment]
@Upabjojr 
Does Travis CI use the updated test files for approving build?

**Comment by Upabjojr** [comment]
@SagarB-97 it is necessary to update all occurrences affected by your edits.

There are instances in the documentation strings, for example. They need to be updated.

Just look at the test errors, they tell you the locations of failures.

> Does Travis CI use the updated test files for approving build?

What is a test file?

**Comment by SagarB-97** [comment]
@Upabjojr 
Travis CI checks are failing because of Examples in Docstrings. (That is Doctests are failing)

I am now updating all docstrings by referring to the error logs.

I wanted to know if Travis CI uses files in my repository or the original repository for performing checks. The checks will pass only if the updated files in my repository are used!

**Comment by Upabjojr** [comment]
Travis checks each commit. Tests for PR are passed if last commit (note: only the last commit) passes.

File: sympy/printing/str.py [hunk 6]
```diff
@@ -74,13 +74,16 @@ def _print_BooleanTrue(self, expr):
...
-        return '%s(%s)' % (expr.func, ', '.join(sorted(self._print(a) for a in
-            expr.args)))
+        return '%s' %(' & '.join([self.parenthesize(arg, 38)
+            for arg in expr.args]))
 
     def _print_Or(self, expr):
-        return '%s(%s)' % (expr.func, ', '.join(sorted(self._print(a) for a in
-            expr.args)))
+        return '%s' %(' | '.join([self.parenthesize(arg, 36)
+            for arg in expr.args]))
```
Comment (Upabjojr) [inline sympy/printing/str.py]: ```return self.stringify(expr.args, " | ", PRECEDENCE["Or"])```

File: sympy/functions/special/bsplines.py [hunk 7]
```diff
@@ -138,14 +141,7 @@ def bspline_basis_set(d, knots, x):
...
     >>> splines
-    [Piecewise((x**2/2, And(x < 1, x >= 0)),
-               (-x**2 + 3*x - 3/2, And(x < 2, x >= 1)),
-               (x**2/2 - 3*x + 9/2, And(x <= 3, x >= 2)),
-               (0, True)),
-     Piecewise((x**2/2 - x + 1/2, And(x < 2, x >= 1)),
-               (-x**2 + 5*x - 11/2, And(x < 3, x >= 2)),
-               (x**2/2 - 4*x + 8, And(x <= 4, x >= 3)),
-               (0, True))]
+    [Piecewise((x**2/2, (x >= 0) & (x < 1)), (-x**2 + 3*x - 3/2, (x >= 1) & (x < 2)), (x**2/2 - 3*x + 9/2, (x >= 2) & (x <= 3)), (0, True)), Piecewise((x**2/2 - x + 1/2, (x >= 1) & (x < 2)), (-x**2 + 5*x - 11/2, (x >= 2) & (x < 3)), (x**2/2 - 4*x + 8, (x >= 3) & (x <= 4)), (0, True))]
```
Comment (Upabjojr) [inline sympy/functions/special/bsplines.py]: maybe we can keep the line feeds, lines in the documentation examples should be shorter than 80 characters, otherwise the HTML documentation on the website will exceed the page border.

File: sympy/functions/special/bsplines.py [hunk 8]
```diff
@@ -43,27 +43,30 @@ def bspline_basis(d, knots, n, x, close=True):
...
-                  (-x**3/6 + 2*x**2 - 8*x + 32/3, And(x <= 4, x >= 3)),
-                  (0, True))
+        Piecewise((x**3/6, (x >= 0) & (x < 1)),
+            (-x**3/2 + 2*x**2 - 2*x + 2/3,
+            (x >= 1) & (x < 2)),
+            (x**3/2 - 4*x**2 + 10*x - 22/3,
+            (x >= 2) & (x < 3)),
+            (-x**3/6 + 2*x**2 - 8*x + 32/3,
+            (x >= 3) & (x <= 4)),
+            (0, True))
```
Comment (Upabjojr) [inline sympy/functions/special/bsplines.py]: keep them aligned with the parenthesis opening.

File: sympy/assumptions/tests/test_query.py [hunk 9]
```diff
@@ -2046,17 +2046,6 @@ def test_compute_known_facts():
...
 
-@slow
-def test_known_facts_consistent():
-    """"Test that ask_generated.py is up-to-date"""
-    from sympy.assumptions.ask import get_known_facts, get_known_facts_keys
-    from os.path import abspath, dirname, join
-    filename = join(dirname(dirname(abspath(__file__))), 'ask_generated.py')
-    with open(filename, 'r') as f:
-        assert f.read() == \
-            compute_known_facts(get_known_facts(), get_known_facts_keys())
```
Comment (Upabjojr) [inline sympy/assumptions/tests/test_query.py]: why has this code been removed?

File: doc/src/modules/logic.rst [hunk 10]
```diff
@@ -18,11 +18,11 @@ You can build Boolean expressions with the standard python operators ``&``
...
     >>> x, y = symbols('x,y')
     >>> y | (x & y)
-    Or(And(x, y), y)
+    y | (x & y)
     >>> x | y
-    Or(x, y)
+    x | y
     >>> ~x
-    Not(x)
+    ~(x)
```
Comment (Upabjojr) [inline doc/src/modules/logic.rst]: this could go without parenthesis, look at the comment to the Not printer.

File: sympy/plotting/plot_implicit.py [hunk 11]
```diff
@@ -283,13 +283,8 @@ def plot_implicit(expr, x_var=None, y_var=None, **kwargs):
...
 
-    >>> p7 = plot_implicit(And(y > x, y > -x))
-
-    When plotting an expression with a single variable (y - 1, for example),
-    specify the x or the y variable explicitly:
-
-    >>> p8 = plot_implicit(y - 1, y_var=y)
-    >>> p9 = plot_implicit(x - 1, x_var=x)
+    >>> p7 = plot_implicit(y - 1, y_var=y)
+    >>> p8 = plot_implicit(x - 1, x_var=x)
```
Comment (SagarB-97) [inline sympy/plotting/plot_implicit.py]: Yea I'll put it back again!

File: sympy/printing/str.py [hunk 12]
```diff
@@ -74,13 +74,16 @@ def _print_BooleanTrue(self, expr):
     def _print_BooleanFalse(self, expr):
         return "False"
 
+    def _print_Not(self, expr):
+        return '~(%s)' %(expr.args[0])
```
Comment (SagarB-97) [inline sympy/printing/str.py]: Thanks! I'll change that

File: sympy/printing/str.py [hunk 13]
```diff
@@ -74,13 +74,16 @@ def _print_BooleanTrue(self, expr):
...
+
     def _print_And(self, expr):
-        return '%s(%s)' % (expr.func, ', '.join(sorted(self._print(a) for a in
-            expr.args)))
+        return '%s' %(' & '.join([self.parenthesize(arg, 38)
```
Comment (SagarB-97) [inline sympy/printing/str.py]: Great! I shall do this!

File: sympy/functions/special/bsplines.py (hunk 7 above)
Comment (SagarB-97) [inline sympy/functions/special/bsplines.py]: Okay I'll change that

File: sympy/assumptions/tests/test_query.py (hunk 9 above)
Comment (SagarB-97) [inline sympy/assumptions/tests/test_query.py]: This line was giving assertion error

```python
assert f.read() == compute_known_facts(get_known_facts(), get_known_facts_keys())
```

It has no Boolean expressions. I wanted to create an Issue for this.

Comment (Upabjojr) [inline sympy/assumptions/tests/test_query.py]: You cannot just remove unrelated code.

Comment (SagarB-97) [inline sympy/assumptions/tests/test_query.py]: Yup. Sorry I'll put it back

Comment (Upabjojr) [inline sympy/assumptions/tests/test_query.py]: Sometimes there are spurious fails in Travis. In that case we just restart the test.

File: sympy/printing/str.py (hunk 13 above)
Comment (SagarB-97) [inline sympy/printing/str.py]: I cannot use `PRECEDENCE["And"]` because `PRECEDENCE["And"]` < `PRECEDENCE["RELATIONAL"]`.

However in python, precedence of `&` is greater than precedence of relational operators.
This means Expressions like : 
```python
And(x<=y,y<=z)
```
will be printed as : 
```python
x <= y & y <= z
```
However, it should be printed as : 
```python
(x <= y) & (y <= z)
```

File: sympy/printing/str.py (hunk 6 above)
Comment (SagarB-97) [inline sympy/printing/str.py]: `PRECEDENCE["Or"]` cannot be used.
(Read Comment for And)

File: sympy/printing/str.py (hunk 13 above)
Comment (Upabjojr) [inline sympy/printing/str.py]: Maybe it's a bug in `PRECEDENCE`. Give me some time.

Comment (SagarB-97) [inline sympy/printing/str.py]: I don't think it's a bug.

In python, precedence of `and` (That is, Boolean and) is less than precedence of relational operators.
However, precedence of `&` (That is, bitwise &) is greater than relational operators.

As `And` objects in SymPy represent Boolean `and` operation, `PRECEDENCE["And"]` should be less than `PRECEDENCE["Relational"]`.
However, we want to represent `And` objects using `&` operator. Therefore, I've used a number.

Comment (Upabjojr) [inline sympy/printing/str.py]: Well, it depends on what you are printing. Different targets have different precedence.

Let's make it simple, let's just use any trick to solve this (i.e. arbitrary value for precedence) and then write a short comment about why you did so.

File: sympy/printing/str.py (hunk 6 above)
Comment (Upabjojr) [inline sympy/printing/str.py]: 36 is OK

Comment (Upabjojr) [inline sympy/printing/str.py]: Add a comment on why you used 36 instead of `PRECEDENCE["Or"]`

File: sympy/printing/str.py (hunk 13 above)
Comment (Upabjojr) [inline sympy/printing/str.py]: I mean... 38 is OK

Comment (SagarB-97) [inline sympy/printing/str.py]: Alright! Thanks :+1:

Comment (SagarB-97) [inline sympy/printing/str.py]: Shall I put in a comment like this : 
```python
# The number 38 has been used (instead of PRECEDENCE["And"]) as 
# Bitwise and operator (&) is used to represent And objects in SymPy
# and not Boolean and (and).
```

File: sympy/printing/str.py (hunk 6 above)
Comment (SagarB-97) [inline sympy/printing/str.py]: Shall I put in a comment like this : 
```python
# The number 36 has been used (instead of PRECEDENCE["Or"]) as 
# Bitwise or operator (|) is used to represent Or objects in SymPy
# and not Boolean or (or).
```

File: sympy/printing/str.py (hunk 13 above)
Comment (Upabjojr) [inline sympy/printing/str.py]: Yes, can go.

**Comment by SagarB-97** [comment]
@Upabjojr 
Review these changes and tell me if they are sufficient.
I shall change the docstrings after the printing format is perfect.

Some examples : 
```python
>>> from sympy import *
>>> x,y,z,a,b,c = symbols('x y z a b c')
>>> And(x,y,z,Or(a,b,c),Not(z))
x & y & z & ~z & (a | b | c)
>>> And(x>=y,y>=z,z>=10,Or(y>=10,x>=10))
(x >= y) & (y >= z) & (z >= 10) & ((x >= 10) | (y >= 10))
>>> Not(y>=10)
y < 10
>>> And(x,y,Not(x>y),z)
x & y & z & (x <= y)
>>> Not(x)
~x
>>> Not(x-y)
~(x - y)
```

**Comment by Upabjojr** [comment]
This looks OK.

**Comment by SagarB-97** [comment]
Alright thanks! I'll start changing the docstrings.

**Comment by SagarB-97** [comment]
@Upabjojr 
There is one error flagged by Travis CI.
It is related to plotting module.

```
File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 78, in test_matplotlib
    plot_implicit_tests('test')
  File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 44, in plot_implicit_tests
    plot_and_save(And(y > exp(x), y > x + 2))
  File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 17, in plot_and_save
    p = plot_implicit(expr, *args, **kwargs)
  File "sympy/plotting/plot_implicit.py", line 369, in plot_implicit
    p.show()
  File "sympy/plotting/plot.py", line 196, in show
    self._backend.show()
  File "sympy/plotting/plot.py", line 1029, in show
    self.process_series()
  File "sympy/plotting/plot.py", line 932, in process_series
    points = s.get_raster()
  File "sympy/plotting/plot_implicit.py", line 80, in get_raster
    temp = func(xinterval, yinterval)
  File "sympy/plotting/experimental_lambdify.py", line 311, in __call__
    return self.lambda_func(*args, **kwargs)
  File "<string>", line 1, in <lambda>
TypeError: unsupported operand type(s) for &: 'tuple' and 'tuple'
```
There are no other errors.
How shall I proceed?

**Comment by SagarB-97** [comment]
@Upabjojr 
Sorry to bug you for such a small patch 😅 
I didn't know it'll involve the laborious task of changing tons of tests and docstrings. 😕

**Comment by Upabjojr** [comment]
> I didn't know it'll involve the laborious task of changing tons of tests and docstrings.

Technically, you could make this task automatic or semi-automatic, once you have all error lines from the tests.

> How shall I proceed?

Did you try debugging the code generating that error?

File: sympy/printing/str.py [hunk 14]
```diff
@@ -74,13 +74,20 @@ def _print_BooleanTrue(self, expr):
...
         return "False"
 
+    def _print_Not(self, expr):
+        return '~%s' %(self.parenthesize(expr.args[0],PRECEDENCE["Not"]))
+
     def _print_And(self, expr):
-        return '%s(%s)' % (expr.func, ', '.join(sorted(self._print(a) for a in
-            expr.args)))
+        return self.stringify(expr.args, " & ", 38)
+    # The number 38 has been used (instead of PRECEDENCE["And"]) as
```
Comment (asmeurer) [inline sympy/printing/str.py]: It would be better to add something to the precedence dict, rather than using a magic number here.

File: sympy/plotting/plot_implicit.py (hunk 11 above)
Comment (asmeurer) [inline sympy/plotting/plot_implicit.py]: How is `str(And(y > x, y > -x))` printed in your branch?

File: sympy/printing/str.py (hunk 14 above)
Comment (Upabjojr) [inline sympy/printing/str.py]: I think this needs some more meditation. Different printers have different precedences, for example the C and Fortran printers would expect `And` to have lower priority than inequalities.

I once create `precedence_traditional` for the LaTeX and pretty printers, maybe we should have something like `precedence_str`.

File: sympy/plotting/plot_implicit.py (hunk 11 above)
Comment (SagarB-97) [inline sympy/plotting/plot_implicit.py]: @asmeurer 
 ```python
(y > x) & (y > -x)
```

File: sympy/printing/str.py (hunk 14 above)
Comment (SagarB-97) [inline sympy/printing/str.py]: @asmeurer @Upabjojr 
So should I create `precedence_str` Dictionary in `str.py` ??

**Comment by Upabjojr** [comment]
> So should I create precedence_str Dictionary in str.py ??

If you feel like, but I would recommend a quicker fix.

Otherwise a quicker fix would be to add 36 and 38 as `BitwiseAnd` and `BitwiseOr` (or vice versa, check) in the `PRECEDENCE` dictionary.

The `precedence_str( )` way could be opened an issue for future correction. I'm not sure whether it's the best way to proceed.

**Comment by SagarB-97** [comment]
@Upabjojr @asmeurer 
All Travis tests except this have passed.
I need some help correcting it.
```python
File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 78, in test_matplotlib
    plot_implicit_tests('test')
  File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 44, in plot_implicit_tests
    plot_and_save(And(y > exp(x), y > x + 2))
  File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 17, in plot_and_save
    p = plot_implicit(expr, *args, **kwargs)
  File "sympy/plotting/plot_implicit.py", line 369, in plot_implicit
    p.show()
  File "sympy/plotting/plot.py", line 196, in show
    self._backend.show()
  File "sympy/plotting/plot.py", line 1029, in show
    self.process_series()
  File "sympy/plotting/plot.py", line 932, in process_series
    points = s.get_raster()
  File "sympy/plotting/plot_implicit.py", line 80, in get_raster
    temp = func(xinterval, yinterval)
  File "sympy/plotting/experimental_lambdify.py", line 311, in __call__
    return self.lambda_func(*args, **kwargs)
  File "<string>", line 1, in <lambda>
TypeError: unsupported operand type(s) for &: 'tuple' and 'tuple'
```

File: sympy/plotting/plot_implicit.py (hunk 11 above)
Comment (SagarB-97) [inline sympy/plotting/plot_implicit.py]: @asmeurer 
Any leads on how to correct this?

**Comment by Upabjojr** [comment]
@SagarB-97 you have messed up the git history. The diff viewer shows a lot of unrelated changes.

Can you fix it? If you're not comfortable with git, just open a new PR and manually copy your changes.

As a rule of thumb, you shouldn't usually merge code or add unrelated edits into your PRs.

**Comment by SagarB-97** [comment]
@Upabjojr 
I just rebased it with the sympy/sympy repo. I thought there might be some new commits that have tests using Boolean functions which might use old printing format!

**Comment by SagarB-97** [comment]
@Upabjojr @asmeurer 
All Travis tests except this have passed.
I need some help correcting it.
```python
File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 78, in test_matplotlib
    plot_implicit_tests('test')
  File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 44, in plot_implicit_tests
    plot_and_save(And(y > exp(x), y > x + 2))
  File "/home/travis/build/sympy/sympy/sympy/plotting/tests/test_plot_implicit.py", line 17, in plot_and_save
    p = plot_implicit(expr, *args, **kwargs)
  File "sympy/plotting/plot_implicit.py", line 369, in plot_implicit
    p.show()
  File "sympy/plotting/plot.py", line 196, in show
    self._backend.show()
  File "sympy/plotting/plot.py", line 1029, in show
    self.process_series()
  File "sympy/plotting/plot.py", line 932, in process_series
    points = s.get_raster()
  File "sympy/plotting/plot_implicit.py", line 80, in get_raster
    temp = func(xinterval, yinterval)
  File "sympy/plotting/experimental_lambdify.py", line 311, in __call__
    return self.lambda_func(*args, **kwargs)
  File "<string>", line 1, in <lambda>
TypeError: unsupported operand type(s) for &: 'tuple' and 'tuple'
```

The result of executing the above Boolean expression is : 
```python
>>> And(y > exp(x), y > x + 2)
(y > exp(x)) & (y > x + 2)
```

The operands are not tuples!

**Comment by Upabjojr** [comment]
Your branch works on my computer.

Concerning Travis: it had some problems recently.

**Comment by SagarB-97** [comment]
@Upabjojr 
So anymore changes required??

**Comment by Upabjojr** [comment]
I think it could be ready. Let's wait for Travis.

**Comment by SagarB-97** [comment]
@Upabjojr 
Travis done. I guess there are some spurious fails.

**Comment by Upabjojr** [comment]
@SagarB-97 I checked it on my computer. The failures are real. Need to investigate.

**Comment by Upabjojr** [comment]
OK, I think I've got the problem. It's a bit of a mess, but we can do it.

This line:
https://github.com/SagarB-97/sympy/blob/8d4e7b3f334f56939c2983f0b740f1546305d73b/sympy/plotting/experimental_lambdify.py#L273

Now prints `&` instead of `And`. Unfortunately `And` was used later to compare Python tuples (which are not SymPy objects and do not support the `&` operator).

Unfortunately the debugger got lost, because the expression that raises the error was later converted to a string, the re-evaluated:
https://github.com/SagarB-97/sympy/blob/8d4e7b3f334f56939c2983f0b740f1546305d73b/sympy/plotting/experimental_lambdify.py#L307

When it is later called:
https://github.com/SagarB-97/sympy/blob/8d4e7b3f334f56939c2983f0b740f1546305d73b/sympy/plotting/experimental_lambdify.py#L311

the debugger is unable to find the point where the exception is raised.

Need to think a bit about a solution. The points where the tuples are returned are found in this file:

https://github.com/SagarB-97/sympy/blob/8d4e7b3f334f56939c2983f0b740f1546305d73b/sympy/plotting/intervalmath/interval_arithmetic.py#L103

One quick idea would be to replace them with SymPy's `Tuple` object, which is a wrapper around `tuple`.

**Comment by Upabjojr** [comment]
We need a full form printer.

**Comment by Upabjojr** [comment]
Or maybe... replacing `str` with `srepr` in 
https://github.com/SagarB-97/sympy/blob/8d4e7b3f334f56939c2983f0b740f1546305d73b/sympy/plotting/experimental_lambdify.py#L273

Maybe this fix is even better.

**Comment by Upabjojr** [comment]
> Or maybe... replacing str with srepr in

OK, it was a bad idea. Try putting `exprstr = exprstr.replace(" & ", " and ").replace(" | ", " or ")` below that line.

And maybe after that line a comment like `# the & and | operators don't work on tuples, see discussion #12108`

**Comment by SagarB-97** [comment]
@Upabjojr 
Thanks a lot!
All tests passing now; Except some assert statement in `sympy/assumptions/tests/test_query.py` which I have not effectively changed!

Do you think it's a spurious fail? Because that error did not show up in some previous commits but did in others.

**Comment by Upabjojr** [comment]
> Do you think it's a spurious fail? Because that error did not show up in some previous commits but did in others.

If it shows up more than in one commit, it is likely not a spurious failure.

**Comment by Upabjojr** [comment]
@SagarB-97 try running the tests locally.

**Comment by SagarB-97** [comment]
@Upabjojr I think I figured it out. I'll fix and commit again.

**Comment by SagarB-97** [comment]
@Upabjojr 
Travis Tests done!!

**Comment by Upabjojr** [comment]
Wait... did you figure out the reason why running the ask update actually solves the problem? It seems a bit strange to me...

**Comment by Upabjojr** [comment]
(before merging I want to make sure that we are not introducing some problems to the development master)

**Comment by SagarB-97** [comment]
Yes.
The error is caused by this assert statement : 

https://github.com/SagarB-97/sympy/blob/5133a321fa46b7f9d828bd6829bb437d4317487f/sympy/assumptions/tests/test_query.py#L2056

`f.read()` returns raw contents of `ask_generated.py`.

`compute_known_facts(get_known_facts(), get_known_facts_keys())` returns value computed by `sympy.assumptions.ask.compute_known_facts`

The `compute_known_facts()` function uses `str()` printer to compute some arguments of `And` : 

https://github.com/SagarB-97/sympy/blob/5133a321fa46b7f9d828bd6829bb437d4317487f/sympy/assumptions/ask.py#L1409

Thus there is a mismatch and the contents of `ask_generated.py` had to be updated.

`sympy/bin/ask_update.py` just replaces the contents of `ask_generated.py` with the string returned by `compute_known_facts : 

https://github.com/SagarB-97/sympy/blob/5133a321fa46b7f9d828bd6829bb437d4317487f/bin/ask_update.py#L28

Even the documentation of `ask_generated.py` reads : 

```python
"""
The contents of this file are the return value of
``sympy.assumptions.ask.compute_known_facts``.

Do NOT manually edit this file.
Instead, run ./bin/ask_update.py.
"""
```

**Comment by Upabjojr** [comment]
This PR is ready. There is still a line with a comma not followed by a space, but that's only a style issue, therefore not important.

@SagarB-97 thanks for your contributions

**Comment by SagarB-97** [comment]
@Upabjojr 
Thank you for guiding me through.
